zenodo/modules/auditor/oai.py

```diff
@@ -0,0 +1,193 @@
+
+        record_oai_sets = (
+            db.session.query(RecordMetadata)
+            .join(
+                PersistentIdentifier,
+                RecordMetadata.id == PersistentIdentifier.object_uuid)
+            .filter(
+                PersistentIdentifier.pid_type == 'recid',
+                PersistentIdentifier.status == PIDStatus.REGISTERED)
+            .with_entities(
+                cast(PersistentIdentifier.pid_value, sa.Integer),
+                RecordMetadata.json)
+        )
+
+        for recid, record in record_oai_sets:
+            for s in record.get('_oai', {}).get('sets', []):
+                self.cache.sadd('{}:{}'.format(self.cache_prefix, s), recid)
+
+    def pop_db_oai_set(self, oai_set):
+        """Return and remove from cache an OAI Set."""
+        key = '{}:{}'.format(self.cache_prefix, oai_set)
+        oai_sets = set(map(int, self.cache.smembers(key)))
+        self.cache.delete(key)
+        return oai_sets
+
+    def clear_db_oai_set_cache(self):
+        """Clear all cached OAI Sets generated for this audit."""
+        for key in self.cache.scan_iter('{}:*'.format(self.cache_prefix)):
+            self.cache.delete(key)
+
+
+class OAICorrespondenceCheck(Check):
+    """OAI Global Check."""
+
+    def sets_correspondence(self):
+        """Check correspondence of Communities and their OAI Sets."""
+        missing_oai_sets = []
+        for c in Community.query.all():
+            if not OAISet.query.filter_by(spec=c.oaiset_spec).count():
+                missing_oai_sets.append(c.id)
+        if missing_oai_sets:
+            self.issues['missing_oai_set'] = missing_oai_sets
+
+        missing_communities = []
+        for s in OAISet.query.filter(OAISet.search_pattern.is_(None),
+                                     OAISet.spec.startswith('user-')):
+            if not Community.query.filter_by(id=s.spec[5:]).count():
+                missing_communities.append(s.spec)
+        if missing_communities:
+            self.issues['missing_community'] = missing_communities
+
+    def perform(self):
+        """Perform OAI Set correspondence checks."""
+        self.sets_correspondence()
+
+
+class OAISetResultCheck(Check):
+    """OAI Set Count check."""
+
+    def __init__(self, audit, community):
+        """Initialize an OAI Set result check."""
+        super(OAISetResultCheck, self).__init__()
+        self.audit = audit
+        self.community = community
+
+    def results_consistency(self):
+        """Check for consistent Community OAI Set results.
+
+        There are three points of reference for OAI Set data:
+          - Database Records
+          - Elasticsearch Records Index
+          - The `/oai2d` Endpoint
+
+        Checking that the set of identifiers is consistent throughout these
+        sources is vital.
+        """
+        db_ids = self._db_identifiers()
+        es_ids = self._es_identifiers()
+        oai2d_ids = self._oai2d_endpoint_identifiers()
+
```

Comment:
S =(es_id | oai2d_ids | db_ids)
for name, X in zip(('oai','es','db', (oai, es, db)):
  self.issues[name] = S - X